google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java

```diff
@@ -149,16 +168,93 @@ public static Builder newBuilder() {
-              .setClientSecret(
-                  properties.getProperty(ConfigPropertyKey.CLIENT_SECRET.getPropertyKey()))
-              .setRefreshToken(
-                  properties.getProperty(ConfigPropertyKey.REFRESH_TOKEN.getPropertyKey()))
-              .build();
-      setCredentials(credentials);
+    private void setCredentials(Properties properties) throws IOException {
+      String serviceAccountSecretsPath =
+          ConfigPropertyKey.SERVICE_ACCOUNT_SECRETS_PATH.getPropertyValue(properties);
+      String refreshToken = ConfigPropertyKey.REFRESH_TOKEN.getPropertyValue(properties);
+      // Validates that entries are present for exactly one of installed app/web flow credentials or
+      // service account credentials.
+      boolean hasInstalledAppKeys =
+          INSTALLED_APP_OAUTH_KEYS.stream().anyMatch(k -> k.getPropertyValue(properties) != null);
+      boolean hasServiceAccountKeys =
+          SERVICE_ACCOUNT_OAUTH_KEYS.stream().anyMatch(k -> k.getPropertyValue(properties) != null);
+      if (!(hasInstalledAppKeys || hasServiceAccountKeys)) {
+        // Entries missing for both types of credentials.
+        throw new IllegalArgumentException(
+            String.format(
+                "Missing properties for credentials. Please modify properties to include entries"
+                    + " for %s if using installed application/web flow credentials, or %s if using"
+                    + " service account credentials.",
+                INSTALLED_APP_OAUTH_KEYS, SERVICE_ACCOUNT_OAUTH_KEYS));
+      } else if (hasInstalledAppKeys && hasServiceAccountKeys) {
+        // Entries specified for both types of credentials.
+        throw new IllegalArgumentException(
+            String.format(
+                "Entries found in properties for both %s and %s. Please modify properties to either"
+                    + " include entries for %s if using installed application/web flow credentials,"
+                    + " or %s if using service account credentials.",
+                ConfigPropertyKey.SERVICE_ACCOUNT_SECRETS_PATH,
+                ConfigPropertyKey.REFRESH_TOKEN,
+                INSTALLED_APP_OAUTH_KEYS,
+                SERVICE_ACCOUNT_OAUTH_KEYS));
+      }
+
+      // Constructs the credentials object using the appropriate set of properties.
+      setCredentials(
+          hasInstalledAppKeys
+              ? createUserCredentials(properties)
+              : createServiceAccountCredentials(properties));
+    }
+
+    /**
+     * Creates and returns a {@link Credentials} object based on the installed app or web server
+     * configuration keys.
+     */
+    private Credentials createUserCredentials(Properties properties) {
+      String refreshToken = ConfigPropertyKey.REFRESH_TOKEN.getPropertyValue(properties);
+      Preconditions.checkNotNull(
+          "No refresh token specified under the key: %s", ConfigPropertyKey.REFRESH_TOKEN);
```

Comment:
This will never fail - `checkNotNull` isn't passed the possibly-null object.

Reply:
Well that's embarrassing....fixed